### ISSUE TITLE
Ask a question about version comparison

### ISSUE THREAD

**Issue by cdwmhcc** [vanowm/slickdealsPlus#2]
https://stackoverflow.com/a/53387532

https://jsfiddle.net/vanowm/p7uvtbor/

I can't comment in stackoverflow, sorry to bother you here.Your idea is great, it solved my problem, thank you very much.

But 1.5-dev should be less than 1.5-beta
But it will return 1.5-dev > 1.5-beta

Your code: 
``` js
const compareVersions = ((prep, l, i, r) => (a, b) =>
{
    a = prep(a);
    b = prep(b);
    l = Math.max(a.length, b.length);
    i = 0;
    r = i;
    while (!r && i < l)
        //convert into integer, uncluding undefined values
      r = ~~a[i] - ~~b[i++];

    return r < 0 ? -1 : (r ? 1 : 0);
})(t => ("" + t)
    // treat non-numerical characters as lower version
    // replacing them with a negative number based on charcode of first character
  .replace(/[^\d.]+/g, c => "." + (c.replace(/[\W_]+/, "").toUpperCase().charCodeAt(0) - 65536) + ".")
    // remove trailing "." and "0" if followed by non-numerical characters (1.0.0b);
  .replace(/(?:\.0+)*(\.-\d+(?:\.\d+)?)\.*$/g, "$1")
    // return array
  .split("."));
```

Modified code:
Added a channel array so that channels can be handled more accurately.
``` ts
const compareVersions = (a: string, b: string) => {
  const prep = (version: string) => {
    const channelOrder = ['dev', 'alpha', 'beta', 'rc']
    const channelRegex = new RegExp(`(?:-|\\s)(${channelOrder.join('|')})|$`, 'gi')
    return version
      .toLowerCase()
      .replace(channelRegex, (_, m) => {
        const position = m ? channelOrder.indexOf(m.toLowerCase()) : channelOrder.length
        return '.' + (position >= 0 ? position + 1 : channelOrder.length + 1)
      })
      .replace(/[^.\d]+/g, '.0.')
      .split('.')
      .map((v) => (v === '' ? 0 : parseInt(v, 10)))
  }

  const v1 = prep(a)
  const v2 = prep(b)
  const length = Math.max(v1.length, v2.length)

  for (let i = 0; i < length; i++) {
    const num1 = i < v1.length ? v1[i]! : 0
    const num2 = i < v2.length ? v2[i]! : 0
    if (num1 !== num2) {
      return num1 < num2 ? -1 : 1
    }
  }
  return 0
}
```

**Comment by vanowm** [comment]
Yeah, it's not perfect, unfortunately there is no "standard" for this type of versioning.
You probably should post your own version there, so others might benefit from it.

**Comment by vanowm** [comment]
Also, `dev` before `beta` is arguable. In my understanding, `dev` and `nightly` (which I'd think are same thing) would be closer to the final than alpha or beta.
Just for fun here is reply from Copilot AI:
>Alpha, Beta, Release Candidate (RC), Stable/Production, Nightly, Dev/Development, Snapshot, Gamma, Delta, EAP (Early Access Program), LTS (Long-Term Support), GA (General Availability).

and from Chat GPT4:
>Pre-Alpha, Alpha, Beta, Beta 2, Release Candidate (RC), RC 2, Dev (Development), Nightly, General Availability (GA), Stable, Patch Release, Minor Update, Major Update, Long Term Support (LTS), Maintenance Release, End of Life (EOL)

**Comment by cdwmhcc** [comment]
> Also, `dev` before `beta` is arguable. In my understanding, `dev` and `nightly` (which I'd think are same thing) would be closer to the final than alpha or beta.

Thank you for your response, @vanowm.

I appreciate the clarification and understand the complexity involved in version comparison, especially when non-standard versioning practices are utilized.

Based on our discussion, I suggested implementing a custom priority system configured to better handle version sequences. I've tried to apply this idea in my code modifications, but I'm still encountering some issues. My depth of understanding in JavaScript might not be sufficient to effectively resolve these.

Also, I believe it would be beneficial to share this comparison function under your answer on Stack Overflow, as you suggested.

Your approach to version comparison is excellent. Thank you very much for your assistance and for sharing your expertise.
